### ISSUE TITLE
Add Lambert W function as torch.special.lambertw

### ISSUE THREAD

**Issue by gmgeorg** [pytorch/pytorch#108948]
### 🚀 The feature, motivation and pitch

Implement the Lambert W function as a core mathematical function available in pytorch.  Suggest to put this in `torch.special.lambertw` to follow pattern in [`scipy.special.lambertw`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.lambertw.html) and [`tfp.special.lambertw`](https://www.tensorflow.org/probability/api_docs/python/tfp/math/lambertw). 

Lambert W function ([wiki](https://en.wikipedia.org/wiki/Lambert_W_function), and [Corless et al, 1996](https://cs.uwaterloo.ca/research/tr/1993/03/W.pdf)) is the solution to the equation `x * exp(x) = z`, ie `W(z)=x` with multiple branches (real and complex).  It has many applications in physics & math and skewed/heavy-tailed distribution modeling; recently this also gained attention in the deep learning literature as a way to add Gaussianization (outlier dampeninig / remove heavy tails) layers ([Goerg 2011](https://projecteuclid.org/journals/annals-of-applied-statistics/volume-5/issue-3/Lambert-W-random-variablesa-new-family-of-generalized-skewed-distributions/10.1214/11-AOAS457.full) & [Goerg 2015](https://www.hindawi.com/journals/tswj/2015/909231/), [Wiese et al, 2019](https://www.tandfonline.com/doi/abs/10.1080/14697688.2020.1730426), [Li 2023](https://openreview.net/pdf?id=OXP9Ns0gnIq), amongst others). 

Adding it here as a core `pytorch.special.lambertw` functionality would open up ability for pytorch developers to implement any advances involving Lambert W functions; currently blocked by lack of implementation compared to other standard frameworks/libraries (Boost C++, TensorFlow, scipy, R):

* Boost C++ implementation: https://www.boost.org/doc/libs/1_69_0/libs/math/doc/html/math_toolkit/lambert_w.html
* `sklearn` like `gaussianize` Transformer: https://github.com/gregversteeg/gaussianize
* `LambertW` R package: https://github.com/gmgeorg/LambertW; based on C++ implementation in `lamW` (https://github.com/cran/lamW/blob/master/src/lambertW.cpp)
* TensorFlow implementation of `lambertw`: https://www.tensorflow.org/probability/api_docs/python/tfp/math/lambertw

Particularly it would unblock implemented Lambert W x F distributions in `torch.distributions.*` (akin to their [TensorFlow Probability `LambertWDistribution` counterpart](https://www.tensorflow.org/probability/api_docs/python/tfp/distributions/LambertWDistribution)) and subsequently in `pyro.ai`, as well as Gaussianization layers.

**Fun fact**: for gradient based computations Lambert W function has the nice property that its derivative can be expressed in terms of the original function value (ie fast computation of derivatives): `W'(z) = W(z) / (z * (1 + W(z))`. See eq (3.2) in https://cs.uwaterloo.ca/research/tr/1993/03/W.pdf

### Alternatives

none; the Lambert W function is a core mathematical function that's missing in `pytorch`. 

### Additional context

Basic implementations of Lambert W already exist here:

https://github.com/thibsej/unbalanced-ot-functionals/blob/13f2203b3993d973f929578085ea458c5c1a7a78/common/torch_lambertw.py

https://github.com/AminJun/BreakingCertifiableDefenses/blob/cc469fa48f7efba21f3584e233c4db0c9a4856c1/RandomizedSmoothing/projected_sinkhorn/lambertw.py

Happy to take a stab at this if such a pure Pythonic `torch` implementation is OK, w/o having to go via `torch._C._special` (following examples above & https://www.tensorflow.org/probability/api_docs/python/tfp/math/lambertw for pattern matching / translation).  Would need some guidance on what constitutes an e2e fully implemented new `torch.special.lambertw` function (gradients, unit tests, ...); best a recent example PR that implemented a different basic core mathematical function.

**Comment by gmgeorg** [comment]
Any updates or guidance here?

**Comment by yu45020** [comment]
@gmgeorg Below is one implementation to get the principal branch with positive inputs. The approximation takes only a few iterations to converge. 

It takes some effort to make it a standard activation function. 

Reference: Iacono, R., Boyd, J.P. New approximations to the principal real-valued branch of the Lambert W-function. Adv Comput Math 43, 1403–1436 (2017). https://doi.org/10.1007/s10444-017-9530-3
```python
def lambertw_approx(x: torch.tensor, max_iter=100, tol=1e-8):
    """
    Analytic approximation for Lambert W function on the principal branch and positive x
    Iacono, R., Boyd, J.P. New approximations to the principal real-valued branch of the Lambert W-function. Adv Comput Math 43, 1403–1436 (2017). https://doi.org/10.1007/s10444-017-9530-3
    :param x: (N,)
    :param max_iter:
    :param tol:
    :return:
    """
    assert x.min() > 0
    # initial guess (equation 17)
    a0 = (1 / (1 + .5 * torch.log(1 + x)))
    # update scheme (equation 12)
    w = torch.log(1 + a0 * x)
    for _ in range(max_iter):
        # update w
        x_w_log = torch.log(x / w)
        wn = x_w_log + (w - x_w_log) / (1 + w)
        if torch.max(torch.abs(wn - w)) < tol:
            break
        else:
            w = wn
    return w

```

Benchmark against `scipy.special.lambertw`. 

Computation time (s):
| number of x | Approximation| Scipy|
|---|-----|-----|
| 1e5| 0.0041 | 0.0330| 
|1e6|0.0529|0.2847| 
|1e8|5.8435 | 19.8514| 

```python
y = np.linspace(1e-5, 1e6, int(1e5))
y_ = torch.tensor(y)
t1 = time.time()
y_a = lambertw_approx(y_)
print(time.time() - t1) 

t2 = time.time()
y_s = scipy.special.lambertw(y).real
print(time.time() - t2)  
```

![benchmark](https://github.com/pytorch/pytorch/assets/28139045/f8856a06-8237-4d79-bc45-42e4065ccf19)

**Comment by gmgeorg** [comment]
Thanks for the reference.

I have an implementation already done that has much lower error rates (1e-15 compared to 1e-8 here), at roughly same runtime (using more accurate approximations than Iacono, R., Boyd, J.P.; see also tensorflow reference)

```
# runtimes
lambertw_approx
0.0057697296142578125
scipy
0.04044055938720703
lambertw_custom [custom implementation I based off the tensorflow implementation]
0.0074346065521240234
```
![image](https://github.com/pytorch/pytorch/assets/48655643/00283d16-1be2-4edf-bb47-73feec6c8609)


The request here was generally to just add this function as a standard repertoire in torch (like in many other existing libraries/frameworks).  Adding it specifically as an activation function can come as follow up; primarily users would benefit for being able to use the `lambertw` function -- just like any other function in `torch`, say, `torch.log` or `torch.exp` .

Happy to share the github repo implementation for reference

**Comment by gmgeorg** [comment]
For reference, made the **torchlambertw** repo public now 
https://github.com/gmgeorg/torchlambertw

I still think all of the available functionality there is just what's missing today in the core torch library.

**Comment by cats256** [comment]
Bump, I think this can be useful

**Comment by knods3k** [comment]
Yea, I could also use this.
